numba_typing/overload_list.py

```diff
@@ -0,0 +1,180 @@
+import numba
+from numba import types
+from numba.extending import overload
+from type_annotations import product_annotations, get_func_annotations
+import typing
+from numba import NumbaDeprecationWarning, NumbaPendingDeprecationWarning
+import warnings
+from numba.typed import List, Dict
+from inspect import getfullargspec
+
+
+def overload_list(orig_func):
+    def overload_inner(ovld_list):
+        def wrapper(*args):
+            func_list = ovld_list()
+            sig_list = []
+            for func in func_list:
+                sig_list.append((product_annotations(
+                    get_func_annotations(func)), func))
+            args_orig_func = getfullargspec(orig_func)
+            values_dict = {name: typ for name, typ in zip(args_orig_func.args, args)}
+            defaults_dict = {}
+            if args_orig_func.defaults:
+                defaults_dict = {name: value for name, value in zip(
+                    args_orig_func.args[::-1], args_orig_func.defaults[::-1])}
+            result = choose_func_by_sig(sig_list, values_dict, defaults_dict)
+
+            if result is None:
+                raise numba.TypingError(f'Unsupported types a={a}, b={b}')
+
+            return result
+
+        return overload(orig_func, strict=False)(wrapper)
+
+    return overload_inner
+
+
+def check_int_type(n_type):
+    return isinstance(n_type, types.Integer)
+
+
+def check_float_type(n_type):
+    return isinstance(n_type, types.Float)
+
+
+def check_bool_type(n_type):
+    return isinstance(n_type, types.Boolean)
+
+
+def check_str_type(n_type):
+    return isinstance(n_type, types.UnicodeType)
+
+
+def check_list_type(self, p_type, n_type):
+    res = isinstance(n_type, types.List) or isinstance(n_type, types.ListType)
+    if isinstance(p_type, type):
+        return res
+    else:
+        return res and self.match(p_type.__args__[0], n_type.dtype)
+
+
+def check_tuple_type(self, p_type, n_type):
+    res = False
+    if isinstance(n_type, types.Tuple):
```

Comment:
Instead of using `n_type.key` you could use `n_type.types` which is defined for both `Tuple` and `UniTuple`, so you can unite both branches:

Something like:
```python
if isinstance(n_type, types.Tuple, types.UniTuple):
    res = True
    for p_val, n_val in zip(p_type.__args__, n_type.types):
        res = res and self.match(p_val, n_val)

return False
```

And btw you need to check that size of `p_type.__args__` and `n_type.types` are the same.
And I believe you don't have tests for the case when they are different.